New blog: Deploy llm-d inference with KAITO RAG on AKS (#5253)
Azure/AKS: blog/_posts/2025-09-05-pair-llmd-and-rag-on-aks.md

|Disaggregated KV cache management|Offloads KV cache across local or remote stores, with orchestration from decode-sidecar and advanced cache eviction policies for efficient memory use|Supports much longer input contexts and multiple concurrent sessions with reduced GPU memory overhead, enabling scalability for large RAG pipelines|

## Let’s get started: KAITO RAGEngine backed by llm-d with P/D Disaggregation

With the [prerequisites](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#prerequisites) covered, this guide will dive into the creation of two distinct but related endpoints:

1. **Inference Endpoint**: an OpenAI API compatible inference service in Kubernetes, created by the llm-d stack. Jump to the [llm-d inference endpoint](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#inference-endpoint) section in our GitHub repository to set it up.

> 1. **RAG Endpoint**: a RAG service provisioned by KAITO with the inference endpoint pointed to step (1) for users to efficiently index and query their documents. Check out the steps in this [RAG endpoint](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#rag-endpoint) GitHub repo section to spin up and verify your RAGEngine workspace, which includes a YAML manifest that looks like:

Reviewer: I think we need to add explicit instructions to deploy the gpu-provisioner otherwise the ragengine-llm-d never deploys.

Reviewer 2: I didn't want to add gpu-provisioner in the blog post because that's one extra thing the reader has to install/uninstall. I'd suggest deploying RAGEngine with existing node(s) + the right instance type and label selector, regardless of node type (CPU or GPU)

Reviewer: Got it. It's just that when I went down the path of following the links and provisioning the infrastructure, there wasn't a step where the user deploys a CPU based node. Either I missed the step, or the CPU VM SKU didn't align with what was already in the cluster. I'll resolve this comment for now and will go through it again now that edits have been made.

> 1. **RAG Endpoint**: a RAG service provisioned by KAITO with the inference endpoint pointed to step (1) for users to efficiently index and query their documents. Check out the steps in this [RAG endpoint](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#rag-endpoint) GitHub repo section to spin up and verify your RAGEngine workspace, which includes a YAML manifest that looks like:

Author: ```suggestion
1. **RAG Endpoint**: a RAG service provisioned by KAITO with the inference endpoint pointed to step (1) for users to efficiently index and query their documents. Check out the steps in this [RAG endpoint](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#rag-endpoint) GitHub cookbook section to spin up and verify your RAGEngine workspace, which includes a YAML manifest that looks like:
```

```bash
apiVersion: kaito.sh/v1alpha1 
kind: RAGEngine 
metadata: 
  name: ragengine-llm-d 
spec: 
  compute: 
    instanceType: "Standard_DS2_v2" 
    labelSelector: 
      matchLabels: 
        node.kubernetes.io/instance-type: Standard_DS2_v2
  embedding:
    local:
      modelID: "BAAI/bge-small-en-v1.5"
  inferenceService:
    url: "http://<inference-url>/v1/chat/completions"
```

> > **Note**: This RAG service only requires general-purpose compute, like the Azure `Standard_DS2_v2` SKU shown above, which often provides a more cost-effective and accessible alternative to the GPU-intensive process of continuous fine-tuning.

Reviewer: I believe Standard_DS2_v4 are more readily available these days over v2 so maybe we should use that instead.

## Practical Example: Indexing and Querying 10-K Filings

Now, we'll pair llm-d inference with KAITO RAGEngine to index the latest SEC 10-K filings of NVIDIA and Berkshire Hathaway in PDFs, allowing us to ask questions and quickly extract key financial and strategic insights.

Investors, analysts, and researchers benefit from this approach by bypassing manual document review and accessing accurate, up-to-date information through natural language queries - all within their Kubernetes cluster!

> Stepping through this [finance example section](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#practical-example-indexing-and-querying-10-k-filings), you can specify the context of the `10-K` filings index with a request like:

Author: ```suggestion
Stepping through this [finance cookbook example](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#practical-example-indexing-and-querying-10-k-filings), you can port-forward the RAGEngine service to access your endpoint locally, and specify the context of the `10-K` filings index as follows:
```

---
title: "Pair llm-d Inference with KAITO RAG Advanced Search to Enhance your AI Workflows"
description: "Accelerate AI-driven discovery on Kubernetes with faster insights, greater accuracy, and scalable performance."
date: 2025-09-05
authors: 
- Ernest Wong
- Sachi Desai
categories:
- ai
- developer
tags:
- llm-d
- kaito
- rag
- faiss
- huggingface
---

## Overview

> In this blog, we'll guide you through setting up an OpenAI API compatible inference endpoint with llm-d and integrating with [RAG](https://kaito-project.github.io/kaito/docs/rag) on AKS, and showcasing its value in a key finance use case: indexing the latest SEC 10-K filings for the two S&P 500 companies and querying them. We’ll also highlight the benefits of llm-d based on its architecture and its synergy with RAG.

Author: ```suggestion
In this blog, we'll guide you through setting up an OpenAI API compatible inference endpoint with [llm-d](https://llm-d.ai/docs/architecture) and integrating with [RAG](https://kaito-project.github.io/kaito/docs/rag) on AKS, and showcasing its value in a key finance use case: indexing the latest SEC 10-K filings for the two S&P 500 companies and querying them. We’ll also highlight the benefits of llm-d based on its architecture and its synergy with RAG.
```

|--|--|--|
|Vector store|Stores text data (documents, FAQs, etc.) in a vector format, or numerical representations of meaning. This allows the system to find relevant pieces of information, even if the user’s question uses different words than the original text (semantic search)|[FAISS](https://faiss.ai/) (Facebook AI Similarity Search) is widely used and is like a memory system that understands meaning and not just keywords|
|Embedding model|Converts text (like a phrase or sentence) into a vector that captures the meaning of the text. Even if the words in different searches don’t match exactly, an embedding model can produce similar vectors that indicate the system sees that they are semantically close|Sentence-BERT (sBERT) and HuggingFace embedding models|
|Retriever + LLM|The retriever finds useful information to help the LLM give a more accurate or update-to-date answer. Together, they make the RAG system flexible and grounded in real, relevant data and not just what the model memorized|[LlamaIndex](https://docs.llamaindex.ai/) and [LangChain](https://python.langchain.com/docs/introduction/) offer open-source retrievers that are useful for different types of data|

> There is where [Kubernetes AI Toolchain Operator](https://kaito-project.github.io/kaito/docs/) (KAITO) RAGEngine brings cloud-native agility to AI application development. By automatically configuring and orchestrating the RAG pipeline on Kubernetes, it lets data scientists and developers focus on building high-impact AI apps, while the engine helps cluster operators and platform engineers handle scaling, rapid iteration, and real-time data grounding.

Author: ```suggestion
There is where [Kubernetes AI Toolchain Operator](https://kaito-project.github.io/kaito/docs/) (KAITO) RAGEngine brings cloud-native agility to AI application development. By automatically configuring and orchestrating the RAG pipeline on Kubernetes, it lets developers focus on building high-impact AI apps, while the engine helps cluster operators and platform engineers handle scaling, rapid iteration, and real-time data grounding.
```

![KAITO RAGEngine Architecture](/assets/images/pair-llmd-and-rag-on-aks/kaito-rag-arch.png)

> KAITO is a CNCF Sandbox project that makes it easy to deploy, serve, and scale LLMs on Kubernetes, without needing to become a DevOps expert.
> 

Author: ```suggestion
```

|Intelligent routing (via Gateway API Inference Extension's Endpoint Picker)|Schedules requests based not only on server queue length and available KV cache size, but also prefix cache hit probability. Maximize the reuse of KV cache for queries with overlapping system prompts and common context retrieved from the vector store|Minimizes latency; enables efficient reuse of KV cache for similar or repeated contexts; handles RAG queries of mixed context length and structure robustly|
|Disaggregated KV cache management|Offloads KV cache across local or remote stores, with orchestration from decode-sidecar and advanced cache eviction policies for efficient memory use|Supports much longer input contexts and multiple concurrent sessions with reduced GPU memory overhead, enabling scalability for large RAG pipelines|

## Let’s get started: KAITO RAGEngine backed by llm-d with P/D Disaggregation

With the [prerequisites](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#prerequisites) covered, this guide will dive into the creation of two distinct but related endpoints:

> 1. **Inference Endpoint**: an OpenAI API compatible inference service in Kubernetes, created by the llm-d stack. Jump to the [llm-d inference endpoint](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#inference-endpoint) section in our GitHub repository to set it up.

Reviewer: Should also leave a note to ensure the reader has access to meta-llama/Llama-3.1-8B-Instruct by visiting https://huggingface.co/meta-llama/Llama-3.1-8B-Instruct to be whitelisted.

1. **RAG Endpoint**: a RAG service provisioned by KAITO with the inference endpoint pointed to step (1) for users to efficiently index and query their documents. Check out the steps in this [RAG endpoint](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#rag-endpoint) GitHub repo section to spin up and verify your RAGEngine workspace, which includes a YAML manifest that looks like:

```bash
apiVersion: kaito.sh/v1alpha1 
kind: RAGEngine 
metadata: 
  name: ragengine-llm-d 
spec: 
  compute: 

>     instanceType: "Standard_DS2_v2" 
>     labelSelector: 
>       matchLabels: 
>         node.kubernetes.io/instance-type: Standard_DS2_v2

Author: ```suggestion
    instanceType: "Standard_DS2_v4" 
    labelSelector: 
      matchLabels: 
        node.kubernetes.io/instance-type: Standard_DS2_v4
```

  embedding:
    local:
      modelID: "BAAI/bge-small-en-v1.5"
  inferenceService:
    url: "http://<inference-url>/v1/chat/completions"
```

> > **Note**: This RAG service only requires general-purpose compute, like the Azure `Standard_DS2_v2` SKU shown above, which often provides a more cost-effective and accessible alternative to the GPU-intensive process of continuous fine-tuning.

Author: ```suggestion
> **Note**: This RAG service only requires general-purpose compute, like the Azure `Standard_DS2_v4` SKU shown above, which often provides a more cost-effective and accessible alternative to the GPU-intensive process of continuous fine-tuning.
```
